gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableFile.java

```diff
@@ -67,6 +80,101 @@
   private List<OwnerAndPermission> ancestorsOwnerAndPermission;
   /** Checksum of the origin file. */
   private byte[] checksum;
+  /** Attributes to be preserved. [rbuUgGpP] */
+  private PreserveAttributes preserve;
+  /** Partition this file belongs to. */
+  private String partition;
+
+  public static Builder builder(FileSystem originFs, FileStatus origin, Path datasetRoot,
+      CopyConfiguration copyConfiguration) {
+
+    Path relativePath = PathUtils.relativizePath(PathUtils.getPathWithoutSchemeAndAuthority(origin.getPath()),
+        PathUtils.getPathWithoutSchemeAndAuthority(datasetRoot));
+    Path targetPath = new Path(copyConfiguration.getTarget(), relativePath);
+
+    return _hiddenBuilder().originFS(originFs).rootPath(datasetRoot).
+        origin(origin).destination(targetPath).relativeDestination(relativePath).
+        preserve(copyConfiguration.getPreserve());
+  }
+
+  public static class Builder {
+
+    private FileSystem originFs;
+    private Path rootPath;
+    private static Cache<Path, OwnerAndPermission> ownerAndPermissionCache;
```

Comment:
No static objects please as these have unpredictable side effects especially if we start reusing containers in YARN. You may need to create a CopyContext object to hold such state.